### PR TITLE
V0.2 embedded optim

### PR DESCRIPTION
**Pull request by ysqyang**
# Description

1. Moved optimizer logic inside LearningModel; 2. refined single-task, multi-task model abstractions; 3. incorporated dueling architecture into DQN based on the refined abstractions.

## Linked issue(s)/Pull request(s)

- [Issue 171](https://github.com/microsoft/maro/issues/171)

## Type of Change

- [ ] Non-breaking bug fix
- [ ] Breaking bug fix
- [x] New feature
- [ ] Test
- [ ] Doc update
- [ ] Docker update

## Related Component

- [ ] Simulation toolkit
- [x] RL toolkit
- [ ] Distributed toolkit

## Has Been Tested

- OS:
  - [ ] Windows
  - [ ] Mac OS
  - [x] Linux
- Python version:
  - [ ] 3.6
  - [x] 3.7
- Key information snapshot(s):
  
## Needs Follow Up Actions

- [ ] New release package
- [ ] New docker image

## Checklist

- [ ] Add/update the related comments
- [ ] Add/update the related tests
- [ ] Add/update the related documentations
- [ ] Update the dependent downstream modules usage

### CODE REVIEW

File: docs/source/examples/cim.rst [hunk 1]
```diff
@@ -152,7 +152,7 @@ the DQN algorithm and an experience pool for each agent.
                 algorithm = DQN(model_dict={"eval": eval_model},
                                 optimizer_opt=(RMSprop, config.agents.algorithm.optimizer),
                                 loss_func_dict={"eval": smooth_l1_loss},
-                                hyper_params=DQNHyperParams(**config.agents.algorithm.hyper_parameters,
+                                hyper_params=DQNConfig(**config.agents.algorithm.hyper_parameters,
```
Comment (ArthurJiang) [inline docs/source/examples/cim.rst:155]: hyper-> config

File: maro/rl/__init__.py [hunk 2]
```diff
@@ -7,7 +7,7 @@
 from maro.rl.agent.abs_agent_manager import AbsAgentManager, AgentManagerMode
 from maro.rl.agent.simple_agent_manager import SimpleAgentManager
 from maro.rl.algorithms.abs_algorithm import AbsAlgorithm
-from maro.rl.algorithms.dqn import DQN, DQNHyperParams
+from maro.rl.algorithms.dqn import DQN, DQNConfig, DuelingDQNTask
```
Comment (ArthurJiang) [inline maro/rl/__init__.py:10]: keep both: hyper and config

File: maro/rl/models/learning_model.py [hunk 3]
```diff
@@ -1,18 +1,45 @@
...
         self._net = nn.Sequential(*block_list)
+        self._is_trainable = optimizer_opt is not None
+        if self._is_trainable:
+            self._optimizer = optimizer_opt[0](self._net.parameters(), **optimizer_opt[1])
+        else:
+            for param in self._net.parameters():
+                param.requires_grad = False
+
+    def __getstate__(self):
+        dic = self.__dict__.copy()
```
Comment (ArthurJiang) [inline maro/rl/models/learning_model.py:31]: state_dict

File: maro/rl/models/abs_learning_model.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    @abstractmethod
+    def step(self, *losses):
+        """Use losses to back-propagate gradients and apply the gradients to the underlying parameters."""
+        return NotImplemented
+
+    def copy(self):
+        return clone(self)
+
+    def load(self, state_dict):
+        self.load_state_dict(state_dict)
```
Comment (ArthurJiang) [inline maro/rl/models/abs_learning_model.py:30]: taret.load(eval.dump)

File: maro/rl/models/learning_model.py [hunk 5]
```diff
@@ -1,18 +1,45 @@
...
     """
-    def __init__(self, block_list: list):
+    def __init__(self, block_list: list, optimizer_opt: tuple = None):
         super().__init__()
         self._net = nn.Sequential(*block_list)
+        self._is_trainable = optimizer_opt is not None
+        if self._is_trainable:
+            self._optimizer = optimizer_opt[0](self._net.parameters(), **optimizer_opt[1])
+        else:
+            for param in self._net.parameters():
```
Comment (ArthurJiang) [inline maro/rl/models/learning_model.py:27]: self._net.eval() ?

File: maro/rl/models/learning_model.py [hunk 6]
```diff
@@ -25,39 +52,146 @@ def forward(self, inputs):
...
 
-    The shared blocks must be chainable, i.e., the output dimension of a block must match the input dimension of
-    its successor. Heads must be provided in the form of keyword arguments. If at least one head is provided, the
-    output of the model will be a dictionary with the names of the heads as keys and the corresponding head outputs
-    as values. Otherwise, the output will be the output of the last block.
+class MultiTaskLearningModel(AbsLearningModel):
+    """NN model that consists of multiple task heads and an optional shared stack.
+
+    Args:
+        task_block_dict (dict): Dictionary of network blocks that perform designated tasks.
```
Comment (ArthurJiang) [inline maro/rl/models/learning_model.py:66]: {"task_name": {
"block_list": []
"optimizer_class": 
"optimizer_param"
} 
}

Comment (ArthurJiang) [reply maro/rl/models/learning_model.py:66]: task_dict, share_xxx

Comment (ArthurJiang) [reply maro/rl/models/learning_model.py:66]: XXXX
blocl_list
optimizer.class
optimizer.param

File: maro/rl/models/learning_model.py [hunk 7]
```diff
@@ -25,39 +52,146 @@ def forward(self, inputs):
...
+            del dic["_shared_optimizer"]
+        if "_head_optimizer_dict" in dic:
+            del dic["_head_optimizer_dict"]
+        return dic
 
-    def forward(self, inputs, key=None):
+    def __setstate__(self, dic: dict):
+        self.__dict__ = dic
+
+    def __getitem__(self, task):
```
Comment (ArthurJiang) [inline maro/rl/models/learning_model.py:125]: task_name: str

File: maro/rl/models/learning_model.py [hunk 8]
```diff
@@ -25,39 +52,146 @@ def forward(self, inputs):
...
+    @has_trainable_heads.setter
+    def has_trainable_heads(self, value: bool):
+        self._has_trainable_heads = value
+
+    @property
+    def is_trainable(self):
+        return self._has_trainable_shared_layers or self._has_trainable_heads
+
+    @property
+    def tasks(self) -> [str]:
```
Comment (ArthurJiang) [inline maro/rl/models/learning_model.py:153]: task_head_names

File: maro/rl/models/learning_model.py [hunk 9]
```diff
@@ -25,39 +52,146 @@ def forward(self, inputs):
...
+    def step(self, loss):
+        """Use the loss to back-propagate gradients and apply them to the underlying parameters."""
+        if not self._has_trainable_shared_layers and not self._has_trainable_heads:
+            raise MissingOptimizerError("No optimizer registered to the model")
+
+        # Zero all gradients
+        if self._has_trainable_shared_layers:
+            self._shared_optimizer.zero_grad()
+        if self._has_trainable_heads:
+            for optim in self._head_optimizer_dict.values():
```
Comment (ArthurJiang) [inline maro/rl/models/learning_model.py:186]: optimizer->task_head_optimizer, str->task_name
net -> task_net

File: maro/rl/algorithms/abs_algorithm.py [hunk 10]
```diff
@@ -3,16 +3,27 @@
...
     algorithms.
+
+    Args:
+        core_model (AbsLearningModel): Task model or container of task models required by the algorithm.
+        config: Settings for the algorithm.
     """
-    def __init__(self):
-        pass
+    def __init__(self, core_model: AbsLearningModel, config):
+        self._core_model = core_model
```
Comment (ArthurJiang) [inline maro/rl/algorithms/abs_algorithm.py:21]: learning_model

File: maro/rl/algorithms/dqn.py [hunk 11]
```diff
@@ -39,97 +67,80 @@ class DQN(AbsAlgorithm):
...
-        self,
-        eval_model: nn.Module,
-        optimizer_cls,
-        optimizer_params,
-        loss_func,
-        hyper_params: DQNHyperParams,
-        target_model=None
-    ):
-        super().__init__()
+    @validate_tasks(DuelingDQNTask)
```
Comment (ArthurJiang) [inline maro/rl/algorithms/dqn.py:73]: validate_task_name

File: maro/rl/models/learning_model.py [hunk 12]
```diff
@@ -25,39 +74,125 @@ def forward(self, inputs):
...
+        return clone(self)
 
-class MultiHeadLearningModel(nn.Module):
-    """NN model that consists of shared blocks and multiple task heads.
 
-    The shared blocks must be chainable, i.e., the output dimension of a block must match the input dimension of
-    its successor. Heads must be provided in the form of keyword arguments. If at least one head is provided, the
-    output of the model will be a dictionary with the names of the heads as keys and the corresponding head outputs
-    as values. Otherwise, the output will be the output of the last block.
+class LearningModel(nn.Module):
```
Comment (ArthurJiang) [inline maro/rl/models/learning_model.py:89]: LearningModuleManager

**Review by wesley-stone — DISMISSED** [review]

**Review by Jinyu-W — APPROVED** [review]

**Pull request merged**